src/Kestrel.Core/Internal/BufferReader.cs

```diff
@@ -1,153 +1,494 @@
+        {
+            ReadOnlySpan<T> remaining = UnreadSpan;
+            int index = delimiters.Length == 2
+                ? remaining.IndexOfAny(delimiters[0], delimiters[1])
+                : remaining.IndexOfAny(delimiters);
+
+            if (index != -1)
+            {
+                span = remaining.Slice(0, index);
+                Advance(index + (advancePastDelimiter ? 1 : 0));
+                return true;
+            }
+
+            return TryReadToAnySlow(out span, delimiters, advancePastDelimiter);
+        }
+
+        private bool TryReadToAnySlow(out ReadOnlySpan<T> span, ReadOnlySpan<T> delimiters, bool advancePastDelimiter)
+        {
+            if (!TryReadToAnyInternal(out ReadOnlySequence<T> sequence, delimiters, advancePastDelimiter, CurrentSpan.Length - CurrentSpanIndex))
+            {
+                span = default;
+                return false;
+            }
+
+            span = sequence.IsSingleSegment ? sequence.First.Span : sequence.ToArray();
+            return true;
+        }
+
+        private bool TryReadToAnyInternal(out ReadOnlySequence<T> sequence, ReadOnlySpan<T> delimiters, bool advancePastDelimiter, int skip = 0)
+        {
+            BufferReader<T> copy = this;
+            if (skip > 0)
+                Advance(skip);
+            ReadOnlySpan<T> remaining = CurrentSpanIndex == 0 ? CurrentSpan : UnreadSpan;
+
+            while (!End)
+            {
+                int index = delimiters.Length == 2
+                    ? remaining.IndexOfAny(delimiters[0], delimiters[1])
+                    : remaining.IndexOfAny(delimiters);
+
+                if (index != -1)
+                {
+                    // Found one of the delimiters. Move to it, slice, then move past it.
+                    if (index > 0)
+                    {
+                        Advance(index);
+                    }
+
+                    sequence = Sequence.Slice(copy.Position, Position);
+                    if (advancePastDelimiter)
+                    {
+                        Advance(1);
+                    }
+                    return true;
+                }
+
+                Advance(remaining.Length);
+                remaining = CurrentSpan;
+            }
+
+            // Didn't find anything, reset our original state.
+            this = copy;
+            sequence = default;
+            return false;
+        }
+
+        /// <summary>
+        /// Check to see if the given <paramref name="next"/> value is next.
+        /// </summary>
+        /// <param name="advancePast">Move past the <paramref name="next"/> value if found.</param>
+        /// <param name="next">The value to look for.</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public bool IsNext(T next, bool advancePast = false)
+        {
+            if (End)
+                return false;
```

Comment:
Does this provide a perf benefit over doing:
```C#
if (!End)
{
   ...
}
return false;
```
?

Reply:
I don't recall, I'll change it corefxlab. The code here is temporary.